doc: add design document for Cluster vs Namespace (#1359)
bpfman/bpfman: docs/design/clusterVsNamespaceScoped.md

## Introduction

For security reasons, cluster admins may want to limit certain applications to only loading eBPF programs
within a given namespace.
Currently, all bpfman Custom Resource Definitions (CRDs) are Cluster scoped.
To provide cluster admins with tighter controls on eBPF program loading, some of the bpfman CRDs also need
to be Namespace scoped.

Not all eBPF programs make sense to be namespaced scoped.
Some eBPF programs like kprobe cannot be constrained to a namespace.
The following programs will have a namespaced scoped variant:

* Uprobe
* TC
* TCX
* XDP

There will also be a namespace scoped BpfApplication variant that is limited to namespaced scoped
eBPF programs listed above.

## Current Implementation

Currently, the reconciler code is broken into two layers (for both the bpfman-operator and the bpfman-agent).
There is the \*Program layer, where there is a reconcile for each program type (Fentry, Fexit, Kprobe, etc).
At this layer, the program specific code handles creating the program specific structure.
The \*Program layer then calls the Common layer to handle processing that is common across all programs.

> 

Author: Done (added `s` to content).

There are some structures, then an interface that defines the set of methods the structure needs to support.

### struct

There are a set of structures (one for the BPF Program CRD and then one for each \*Program CRD) that define the
content of the CRDs (bpfman-operator/apis/v1alpha).
Each object (BPF Program CRD and \*Program CRD) also has a List object.

```go
type BpfProgram struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec BpfProgramSpec `json:"spec"`
	// +optional
	Status BpfProgramStatus `json:"status,omitempty"`
}
type BpfProgramList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []BpfProgram `json:"items"`
}

type FentryProgram struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec FentryProgramSpec `json:"spec"`
	// +optional
	Status FentryProgramStatus `json:"status,omitempty"`
}
type FentryProgramList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []FentryProgram `json:"items"`
}

> :

Author: Done (removed `in` from in front of `each`).

	getLoadRequest(bpfProgram *bpfmaniov1alpha1.BpfProgram,
    mapOwnerId *uint32) (*gobpfman.LoadRequest, error)
	getNode() *v1.Node
	getBpfProgramCommon() *bpfmaniov1alpha1.BpfProgramCommon
	setCurrentProgram(program client.Object) error
	getNodeSelector() *metav1.LabelSelector
	getBpfGlobalData() map[string][]byte
	getAppProgramId() string
}
```

There are also some common reconciler functions that perform common code.

```go
func (r *ReconcilerCommon) reconcileCommon(ctx context.Context,
  rec bpfmanReconciler,
	programs []client.Object) (bool, ctrl.Result, error) {
:
}

func (r *ReconcilerCommon) reconcileBpfProgram(ctx context.Context,
	rec bpfmanReconciler,
	loadedBpfPrograms map[string]*gobpfman.ListResponse_ListResult,
	bpfProgram *bpfmaniov1alpha1.BpfProgram,
	isNodeSelected bool,
	isBeingDeleted bool,
	mapOwnerStatus *MapOwnerParamStatus)
(bpfmaniov1alpha1.BpfProgramConditionType, error) {
:
}

func (r *ReconcilerCommon) reconcileBpfProgramSuccessCondition(
	isLoaded bool,
	shouldBeLoaded bool,
	isNodeSelected bool,
	isBeingDeleted bool,
	noContainersOnNode bool,
  mapOwnerStatus *MapOwnerParamStatus) bpfmaniov1alpha1.BpfProgramConditionType {
:
}
```

> So looks something like this:

Reviewer: ```suggestion
## Adding Namespaced Scoped CRDs
```

Author: Done (added `CRDs`)

> ```
>                      --- FentryProgramReconciler
>                      |     func (r *FentryProgramReconciler) getFinalizer() string {}

Reviewer: ```suggestion
While the contents are the same for the namespace and cluster-scoped CRD in most cases, Kubernetes requires different 
CRD for each type.
```

Author: Done (replaced `To add Namespace scoped BPF Program and \*Program CRDs, Kubernetes requires new CRDs for each type.` with your text. Did add the word `mostly` the same, since we had to remove namespace from ContainerSelector.)

	// +optional
	Status BpfProgramStatus `json:"status,omitempty"`
}
type BpfProgramList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []BpfProgram `json:"items"`
}

type BpfNsProgram struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec BpfProgramSpec `json:"spec"`
	// +optional
	Status BpfProgramStatus `json:"status,omitempty"`
}
type BpfNsProgramList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []BpfProgram `json:"items"`
}

type TcProgram struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec TcProgramSpec `json:"spec"`
	// +optional
	Status TcProgramStatus `json:"status,omitempty"`
}
type TcProgramList struct {
	metav1.TypeMeta `json:",inline"`
	metav1.ListMeta `json:"metadata,omitempty"`
	Items           []TcProgram `json:"items"`
}

type TcNsProgram struct {
	metav1.TypeMeta   `json:",inline"`
	metav1.ObjectMeta `json:"metadata,omitempty"`

	Spec TcNsProgramSpec `json:"spec"`
	// +optional
	Status TcProgramStatus `json:"status,omitempty"`
}

> type TcNsProgramList struct {

Reviewer: ```suggestion
To allow the common code to act on both Cluster or Namespaced objects, two new interfaces will be introduced.
```

Author: Done (added `d` to `introduce`.)

}
```

Several functions were added to the bpfmanReconciler interface that are implemented by these structures.

```go
type bpfmanReconciler[T BpfProg, TL BpfProgList[T]] interface {
	// BPF Cluster of Namespaced Reconciler
	getBpfList(ctx context.Context, opts []client.ListOption) (*TL, error)
	updateBpfStatus(ctx context.Context, bpfProgram *T, condition metav1.Condition) error
	createBpfProgram(
		attachPoint string,
		rec bpfmanReconciler[T, TL],
		annotations map[string]string,
	) (*T, error)

	// *Program Reconciler
  SetupWithManager(mgr ctrl.Manager) error
	:
}
```

And the \*Programs use the `ClusterProgramReconciler` or `NamespaceProgramReconciler` structs instead
of the `ReconcilerCommon` struct.

```go
type TcProgramReconciler struct {
	ClusterProgramReconciler
	currentTcProgram *bpfmaniov1alpha1.TcProgram
	interfaces       []string
	ourNode          *v1.Node
}

type TcNsProgramReconciler struct {
	NamespaceProgramReconciler
	currentTcNsProgram *bpfmaniov1alpha1.TcNsProgram
	interfaces         []string
	ourNode            *v1.Node
}

:
```

## Naming

In the existing codebase, all the CRDs are cluster scoped:

* BpfApplicationProgram
* BpfProgram

> * FentryProgram

Reviewer: ```suggestion
Around the same time this feature is being developed, another feature is being developed
```

Author: Done (removed `r` from `featurer`)

* FexitProgram
* KprobeProgram
* ...

> 
> Common practice is for cluster scoped objects to include "Cluster" in the name and
> for namespaced objects to not have an identifier.

Reviewer: ```suggestion
carry the identifier.  After the load/attach split work is complete, the CRDs will be renamed to follow the common convention in which the cluster-scoped CRD names are prefixed with "Cluster".
```

We may not do a v1alpha2 now.

Author: Done
